beacon/sync/src/test/java/tech/pegasys/teku/beacon/sync/historical/ReconstructHistoricalStatesServiceTest.java

```diff
@@ -0,0 +1,173 @@
+import tech.pegasys.teku.infrastructure.unsigned.UInt64;
+import tech.pegasys.teku.spec.Spec;
+import tech.pegasys.teku.spec.TestSpecFactory;
+import tech.pegasys.teku.spec.datastructures.state.Checkpoint;
+import tech.pegasys.teku.spec.datastructures.state.beaconstate.BeaconState;
+import tech.pegasys.teku.spec.generator.ChainBuilder;
+import tech.pegasys.teku.storage.api.StorageUpdateChannel;
+import tech.pegasys.teku.storage.client.CombinedChainDataClient;
+
+public class ReconstructHistoricalStatesServiceTest {
+  private final Spec spec = TestSpecFactory.createDefault();
+  private final StorageUpdateChannel storageUpdateChannel = mock(StorageUpdateChannel.class);
+  private final CombinedChainDataClient chainDataClient = mock(CombinedChainDataClient.class);
+  private final ChainBuilder chainBuilder = ChainBuilder.create(spec);
+  private ReconstructHistoricalStatesService service;
+  private final StatusLogger statusLogger = mock(StatusLogger.class);
+
+  @BeforeEach
+  void setup() {
+    chainBuilder.generateGenesis();
+    chainBuilder.generateBlocksUpToSlot(10);
+
+    when(storageUpdateChannel.onFinalizedState(any())).thenReturn(SafeFuture.COMPLETE);
+  }
+
+  @Test
+  public void shouldCompleteExceptionallyWithEmptyResource() {
+    createService(Optional.empty());
+
+    final SafeFuture<?> res = service.start();
+    assertThat(res).isCompletedExceptionally();
+    assertThatSafeFuture(res).isCompletedExceptionallyWith(IllegalStateException.class);
+    assertThatSafeFuture(res)
+        .isCompletedExceptionallyWithMessage("Genesis state resource not provided");
+
+    verify(chainDataClient, never()).getInitialAnchor();
+    verify(storageUpdateChannel, never()).onFinalizedState(any());
+  }
+
+  @Test
+  public void shouldCompleteExceptionallyWithFailedLoadState() {
+    createService(Optional.of("invalid resource"));
+
+    final SafeFuture<?> res = service.start();
+    assertThat(res).isCompletedExceptionally();
+    assertThatSafeFuture(res).isCompletedExceptionallyWith(InvalidConfigurationException.class);
+    assertThatSafeFuture(res)
+        .isCompletedExceptionallyWithMessage(
+            "Failed to load initial state from invalid resource: Not found");
+
+    verify(chainDataClient, never()).getInitialAnchor();
+    verify(storageUpdateChannel, never()).onFinalizedState(any());
+  }
+
+  @Test
+  public void shouldCompleteSkipWithEmptyCheckpoint(@TempDir final Path tempDir)
+      throws IOException {
+    createService(createGenesisStateResource(tempDir));
+    when(chainDataClient.getInitialAnchor())
+        .thenReturn(SafeFuture.completedFuture(Optional.empty()));
+
+    final SafeFuture<?> res = service.start();
+    assertThat(res).isCompleted();
+    verify(chainDataClient, times(1)).getInitialAnchor();
+    verify(storageUpdateChannel, never()).onFinalizedState(any());
+  }
+
+  @Test
+  public void shouldCompleteStart(@TempDir final Path tempDir) throws IOException {
```

Comment:
nit: Probably rename this to `shouldRegenerateStates` - we're testing much more than just that the future from start completes.